src/services/passport.ts

```diff
@@ -58,6 +60,105 @@ const initPassport = () => {
 			},
 		),
 	)
+
+	/**
+	 * Sign in with Google
+	 *
+	 * User has signed in
+	 *  - link user's account with Google if it has not been linked
+	 * User has not signed in
+	 *  - sign in user if google account has been used to sign up
+	 * 	- sign up user if google account has not been used to sign up
+	 */
+	passport.use(
+		new GoogleStrategy(
+			{
+				clientID: config.secrets.googleClientId,
+				clientSecret: config.secrets.googleClientSecret,
+				callbackURL: '/auth/google/callback',
+				passReqToCallback: true,
+			},
+			async (req, accessToken, refreshToken, profile, done) => {
+				const {
+					id: googleId,
+					name: {givenName, familyName},
+					emails: [{value: gmail}],
+				} = profile
+
+				try {
+					if (req.user) {
+						// If google account has been linked, throw error
+						const existingUser = await UserModel.findOne({
+							googleId: profile.id,
+						}).exec()
+
+						if (existingUser) {
+							return done(
+								apiError.badRequest(
+									'Google account has been linked',
+									ErrorCode.googleAccountHasBeenLinked,
+								),
+								false,
+							)
+						}
+
+						// Link google account
+						const user = await UserModel.findById(req.user.id).exec()
+
+						user.googleId = googleId
+						user.firstName = givenName
+						user.lastName = familyName
+
+						const {email} = user
+						if (!email) {
+							user.email = gmail
+						}
+
+						const savedUser = await user.save()
+						return done(null, savedUser)
+					} else {
+						// If google account has been used to sign up, sign in user
+						const existingUser = await UserModel.findOne({
+							googleId: profile.id,
+						}).exec()
+
+						if (existingUser) {
+							return done(null, existingUser)
+						}
+
+						// Create new Google account, if email has been used to sign up, throw error
+						const existingEmailUser = await UserModel.findOne({
+							email: profile.emails[0].value,
+						})
+
+						if (existingEmailUser) {
+							return done(
+								apiError.badRequest(
+									'There is already an account using this Google email',
+									ErrorCode.googleAccountHasBeenTaken,
+								),
+								false,
+							)
+						}
+
+						const newUser: User = {
+							firstName: givenName,
+							lastName: familyName,
+							email: gmail,
+							googleId,
+							status: UserStatus.Active,
+						}
+
+						const createdUser = await UserModel.create(newUser)
```

Comment:
Should we have this user creation done in user service layer?

Reply:
We will address this issue next meeting